### DISCUSSION TITLE
IMU implementation

### DISCUSSION THREAD

**Discussion by YohanBorda** [fjp/2wd-robot#36] [Ideas]
Hi Guys, 
First that all thank for sharing all this information, this facilitated one project that we are buidling for an autonomous vehicule for the agriculture industry. I managed to implement the IMU to your hardware interface using the BOSH BNO055. I'm keen to share this with you in any case you are looking to improve the odometry by adding at ekf filter. 

Just to give you some context in the hardware that we are using, we have the arduino Mega doing all the MCU stuff and running ROS Melodic in a Jetson AGX Xavier.

Just let me know if you are interested in the IMU side in case you haven't done it already.

Update: sorry I just realized that you have already open for this.

**Comment by fjp** [comment]
Hi Yohan, thanks for sharing, I am glad this project is useful for you. It would be awesome if you could share your implementation of the IMU as a new PR. I am happy to review and test it on my robot as soon as I find time. Currently I am working on a new release, where only the wheel encoders are used for odometry. After that I have planned to implement the IMU (#21, #4). So your idea comes at the right time, please feel free to open a PR :-)

Regarding the IMU implementation, I would be really interested to see where you interfaced the IMU: 
- to the "high-level" hardware interface code running on the Jetson AGX Xavier or 
- to the "low-level" Arduino Mega MCU (similar to [Linorobot](https://github.com/linorobot/linorobot/blob/53bc4abf150632fbc6c28cdfcc207caa0f81d2b1/teensy/firmware/src/firmware.ino#L215), [Odometry Wiki](https://github.com/linorobot/linorobot/wiki/3.-Odometry)). 
 
Either way you did it is fine, as I plan to test both. Also which ekf package you used 

In case you used available open source ROS packages for your implementation of the BOSCH BNO055 (e.g. driver node and [ekf packages](https://github.com/ros-planning/robot_pose_ekf)/[robot localization](https://github.com/cra-ros-pkg/robot_localization)), it would be great if you can share this in your PR (regarding licensing and to improve the documentation). Also feel free to share other suggestions about helpful IMU resources (packages, papers, articles, etc), for example about the calibration procedure are most welcome. This would be useful to improve the [documentation](https://ros-mobile-robots.com). However, depending on the time you have for this, just sharing your implementation in a new PR is also much appreciated.
